### ISSUE TITLE
`no-parent-child-coupling` should not flag props getters

### ISSUE THREAD

**Issue by kurtextrem** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#19]
hey, this is less of an issue, I think, and more of a discussion; but the repo doesn't have discussions, so hence opening it as issue :) 

I have a component like this:
```jsx
function Editor({ tool }) {
    const editorRef = useRef()

    const handleBeforeInput = useCallback(
        (event: InputEvent) => {
            if (event.isComposing) return

            const variableInfo = tool.getVariableInfo()
            //                  ^ no error here
            if (variableInfo?.variableDefinition.options?.displayTextArea) return

            e.preventDefault()
        },
        [tool]
    )

    useEffect(() => {
        const editor = editorRef.current
        assert(editor, "Editor must exist")

        const controlProp = tool.getControlProp()
        //                  ^ Avoid coupling parent behavior or state to a child component. Instead, lift shared logic or state up to the parent
        assert(controlProp, "Control prop must exist")

        const value = controlProp.value
        assert(isString(value), "Value must be a string")

        editor.innerText = value
    }, [tool])

    useEffect(() => {
        const editor = editorRef.current
        editor.addEventListener("beforeinput", handleBeforeInput)
        return () => {
            editor.removeEventListener("beforeinput", handleBeforeInput)
        }
    }, [handleBeforeInput])

   return <div ref={editorRef} />
}
```

The error _"Avoid coupling parent behavior or state to a child component. Instead, lift shared logic or state up to the parent"_ is really not easily actionable (hence asking for options here https://github.com/NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect/issues/15#issuecomment-3018050942), but what I do wonder: Do we / you think it would be an improvement here to pass down `getControlProp` instead of `tool` as prop?
IMO it could be argued about, to me that sounds more like a subjective difference? What would be the advantage of doing so? Or can you think of any other fixes?

I can see that it couples `Editor` more tightly to the parent, but is that always a good thing you want to have? Unsure, especially considering the usage in `useCallback` doesn't error (as the lint rule is only about effects).

FWIW, I think making the error more actionable would also be an improvement. Show the message _"Avoid coupling parent behavior or state to a child component. Instead, lift shared logic or state up to the parent"_ to some non-react-pro together with the line and ask them what to do. I'd have no idea. An improvement might be _"Avoid coupling parent behavior or state to a child component. Instead, lift pass down `getControlProp` as prop."_ or so.

**Comment by NickvanDyke** [comment]
> Do we / you think it would be an improvement here to pass down getControlProp instead of tool as prop?

Hmm, your example is flagged because I assume that all method calls on prop objects are callbacks that communicate child -> parent. I didn't consider that it could be a getter method.

In terms of readability, it doesn't seem much different to me.

However, I think the getter breaks React's dependency arrays. Atm your example will always recreate `handleBeforeInput`, and re-run your effect, because `tool` is compared by reference. You could memoize `tool` but there's no way to know whether the getter's would-be result has changed, so then your effect will "under-execute". Ideally you should call the getter outside the effect - eliminating this warning - so you can pass its result to the dependency array.

> An improvement might be "Avoid coupling parent behavior or state to a child component. Instead, lift pass down getControlProp as prop." or so.

Definitely agree - the code has a TODO note for this already, but I know that's not very visible 😀

**Comment by kurtextrem** [comment]
Yeah, the last effect could be replaced by a ref callback, but that might not be a too big issue. 

I thought about this too, but running side-effects in-render looks sketchy (it's not setting state, it's manipulating the DOM!) and the result would be wrong too as react then overrides the DOM after render during commit. 
In this simple case you could set the desired innerText as state, but in reality it's an external system (some kind of editor lib), so I'm pretty sure this is valid use of an effect - it's synchronizing the component with an external system. Thoughts?

**Comment by NickvanDyke** [comment]
It occurred to me that we could check whether the `CallExpression` result is used, to discern these "getter props" from "callback props". Needs further exploration. It's probably not reliable, because ultimately there's no way to know what the prop function does. Even a getter could have side-effects.

I think this is low-priority because the warning is avoidable by calling the getter outside the effect, which imo is more idiomatic and readable anyway. But I'm open to being proved wrong!

**Comment by kurtextrem** [comment]
Awesome improvement. Thank you!
